ddtrace/pin.py

```diff
@@ -104,25 +117,33 @@ def onto(self, obj, send=True):
                 return obj.__setddpin__(self)
 
             pin_name = _DD_PIN_PROXY_NAME if isinstance(obj, wrapt.ObjectProxy) else _DD_PIN_NAME
+
+            # set the target reference; any get_from, clones and retarget the new PIN
+            self._target = obj
             return setattr(obj, pin_name, self)
         except AttributeError:
             log.debug("can't pin onto object. skipping", exc_info=True)
 
     def clone(self, service=None, app=None, app_type=None, tags=None, tracer=None):
-        """ Return a clone of the pin with the given attributes replaced. """
+        """Return a clone of the pin with the given attributes replaced."""
+        # do a shallow copy of Pin dicts
         if not tags and self.tags:
-            # do a shallow copy of the tags if needed.
-            tags = {k:v for k, v in self.tags.items()}
+            tags = self.tags.copy()
+
+        config = self._config.copy()
 
         return Pin(
             service=service or self.service,
             app=app or self.app,
             app_type=app_type or self.app_type,
             tags=tags,
-            tracer=tracer or self.tracer) # no copy of the tracer
+            tracer=tracer or self.tracer,  # do not clone the Tracer
+            _config=config,
```

Comment:
never expose `config` to users from the `Pin` class.

Reply:
Is putting this underscore `_config` hiding it from the public API?
Technically it is still part of the `__init__` which is publicly exposed ; I'm afraid it mislead people / allow bad things.

Reply:
yes it's a convention to hide attributes and elements that are prefixed with an underscore. Of course anyone can use it, but if it has an underscore it's considered extremely dangerous. All frameworks do that, though because of the nature of Python, anyone can access to any attribute that is prefixed.